### ISSUE TITLE
[ORM] [2012-08-23] singleton(<DataObject>) calls populateDefaults() - regression?

### ISSUE THREAD

**Issue by silverstripe-issues** [silverstripe/silverstripe-framework#1575]
created by: @chillu (ischommer)
created at: 2012-08-23
original ticket: http://open.silverstripe.org/ticket/7808

---

singleton() in 2.4 passed the constructor args "null, true" to any object it created.
For most of them, this wouldn't have any effect, but for DataObject the second arg
equated to the $isSingleton boolean. When the flag was set, it didn't call populateDefaults().
So admittedly hacky, but has been there since the dawn of time.

With the introduction of the Injector API in 3.0, this behaviour changed,
no constructor args are passed any longer, meaning poplateDefaults() is called on ALL DataObject singletons.
The problem was uncovered by a modification to how Translatable creates SiteConfig records, see https://github.com/silverstripe/silverstripe-cms/pull/181.

In general, I don't see a problem with calling populateDefaults() on a singleton,
although we'd need to mark it as an API change.
populateDefaults() shouldn't cause any state changing actions beyond
its own in-memory object state, so that's a problem needing to be fixed in the Translatable pull request.

Hamish, Sam, Marcus - do you agree?

**Comment by sminnee** [comment]
Given that this also causes #4878 we should probably address this sooner rather than later (an ironic thing to say given the age of this ticket).

I think what we want is to pass a factory to injector to use, by default, for service/singleton instantiation of any DataObject subclass. This can pass `true` to the 2nd DataObject parameter without this requiring special Injector hard-coding.

Without changes to Injector's capabilities, this would require a separate service definition for every class ("yuck!"), so we need to determine the approach.

Right now Injector has two points at which a factory can be specified:

 * The default factory (Injector::setObjectCreator, which defaults to an InjectionCreator)
 * The factory specified in the service spec's `factory` property

Right now Injector has two points at which a factory can be specified:

 * The default factory (Injector::setObjectCreator, which defaults to an InjectionCreator)
 * The factory specified in the service spec's `factory` property

Perhaps we should elaborate on this with `Injector._defaultFactory` as a config option:
```
Injector:
  _defaultFactory:
    SilverStripe\ORM\DataObject.singleton: DataObjectCreator
    _default: InjectionCreator
```

The _defaultFactory list would be iterated through in the order specified and the service used:
 * Exact class match
 * Subclass-of
 * Implementor-of

...with ".singleton" or ".prototype" as a sub-specifier so that you can limit the setting.

To complete the nesting of turtles, values would in turn be the names of services. ;-)

`Injector._defaultFactory._default` would replace `setObjectCreator`.

cc @nyeholt @silverstripe/core-team 

Alternatively we could have SilverStripe/ORM provide an alternative setObjectCreator value that simply had an "if subclassof DataObject" check baked into it InjectionCreator. It seems like counterproductive coupling, but perhaps it's better than increasing Injector's API surface as described above.

**Comment by sminnee** [comment]
I believe that this bug - the fact that it calls populateDefaults() on singletons as part of DataObject::requireTable() in dev/build - may be the reason that dev/build will sometime fail when migrating a codebase to a new, empty server. As such, bumping priority, which may also warrant backporting the fix to 3.7 and/or 4.1/2

/cc @stojg @jakedaleweb @chillu @mateusz

**Comment by sminnee** [comment]
Having thought about this more I think that a hackier fix that minimises the surface of new APIs created is probably best.

Specifically, have a special case in InjectionCreator for DataObject and its subclasses. It creates coupling between framework-core and framework-orm, but the amount of new API intended to support this one use-case in a less hacky way seems disproportionate.

**Comment by sminnee** [comment]
OK, fun fact, this has already been fixed in February: c767e472dc494408460ef47c27b8d34475da4ac6

It's available in the following stable releases
 * 4.2.1 4.2.0
 * 4.1.2 4.1.1 4.1.0
 * 3.7.1 3.7.0

**Comment by sminnee** [comment]
Futher references, its this PR: https://github.com/silverstripe/silverstripe-framework/pull/7850

It's also in 3.5.x-dev and 3.6.x-dev, but there are no stable tags for those.
